### DISCUSSION TITLE
Blocking chart is not displayed correctly for PDF

### DISCUSSION THREAD

**Discussion by aalexandersson** [moj-analytical-services/splink#1990] [Q&A] [answered]
How can I display the [cumulative_num_comparisons_from_blocking_rules_chart](https://moj-analytical-services.github.io/splink/charts/cumulative_num_comparisons_from_blocking_rules_chart.html) as PNG **for PDF** correctly? I am using Quarto.

On my test data, this code produces the PNG chart for HTML:
`linker.cumulative_num_comparisons_from_blocking_rules_chart()`

The blocking chart is displayed correctly for HTML:
![blocking_chart_html2](https://github.com/moj-analytical-services/splink/assets/20467610/e284fad0-d529-4062-9892-29318ee399ec)

To save the PNG for PDF, I used a basic variation of the recommendation on the Splink [July 2023 blog on how to save Altair 5 charts](https://moj-analytical-services.github.io/splink/blog/2023/07/27/splink-updates---july-2023.html#charts-in-altair-5):
```
block_chart = linker.cumulative_num_comparisons_from_blocking_rules_chart()
block_chart.save("block.png")
```
However, the blocking chart is NOT displayed correctly for PDF. The PNG for PDF has unwanted space to the left of the y-title and, therefore, the rest of the chart is too small:
![block](https://github.com/moj-analytical-services/splink/assets/20467610/21dcc18f-ea58-42f1-a2a0-2b6fa340e53f)

Please advice. Is this a bug in Splink? The issue is only for displaying the blocking chart for PDF (not for HTML, and not for other charts). 

Update 1: The original PNG chart for HTML was outdated, using slightly different data. This has been corrected. The original discussion is [pull 660](https://github.com/moj-analytical-services/splink/pull/660). I am using Splink 3.9.12 (latest) and Altair 5.2.0 (latest). The [Splink documentation on Altair charts](https://moj-analytical-services.github.io/splink/dev_guides/charts/understanding_and_editing_charts.html#charts-in-splink) provides background context. My work-around solution in Quarto is to reuse the PNG graph rendered for HTML as the PNG graph for PDF. It works but is clumsy to have to run code twice. There should be a way to save the PNG chart without distortions when rendering to PDF.

Update 2: The Altair web page for saving charts is [saving_charts.html](https://altair-viz.github.io/user_guide/saving_charts.html). The code `block_chart.to_url()` creates a [sharable URL](https://vega.github.io/editor/#/url/vega-lite/N4Igxg9gdgZglgcxALlANzgUwO4tJKAFzigFcJSBnAdTgBNCALFAZgAY2AacaYsiygAlMiRoVYcAvpO50AhoTl4QUOQFtMKEPMUBaAKwAWAOwBGQwCYAHADYwcuSwBGATkxOrLm4dNgb+lzorKzBTCxsQGRA1OQAnAGstJziQbkwoSDoSJFQeABsIWOV4TDy6LVjSPM1uaoR08uQyPLzuSntq5XbGTA0te0JMBEKATws2MEiowrpMItySsv7SNSqFODRMAH1YiGxKSO5CCAg84gAHFABtUEXGkErOo7hCTuQQAGUARQAZAAIAMLQLLEaCpECEEbnTTvKAQNQkOR5Q63LBLd67bBbSCkIjgmCFGLid7cZ6vGEgIFqc5xOCUaCUP4AcXScwUmHKRyhFIAjqQ5EQXutNiiQHdlqs8sLtpiDtwCbEiVpSRCXm9KSs1sRNoD4TTYnSGeDIdCtHyBcRFNrNDJUaV7vZYoNKHABfjCQplca1RSAXFna6oH8AAq7OikMCEP4QGB-ACSUHOpCjABEFEouab3ubBVaNjbOHb0Q8ORHQVAdusIN7yVoAEqlyNwaB-OtVv4ACjAmql1r+soA9I6AwK-ucw2WAJTG7laOEI1TIyQAXW4AA9imj7pRFE6a+qqfrDVBGSyoGzBnQ-k4Rq2qpgO5Rp5nefzc9LDiA1xZN-aJVr8x2PYDiiEZf2LR4ahAek92QK4QF0b8QFXVVa3eb5-gAIQKMB4myO9OmkbgelEYlQB3TBLmQcYoktN5QEGNdiUpCgiGjWMAEE6BBZtF11alaXpE9mVZWJ2SvG8-gAUTkMBGD+bCIFw-C63vcFKFIJw6IpDsgVxQhGUwNcwDyUhZj+SABINITGSRWJMDkOhb3qc8xMva9b3HTAMAEft7yfT9sHoJgUEMfQuBAAASbpeiUd4xEIc5KGQAcB02BA5AAOgQF5GE0zLmwHGKYjSoY5F0PIXkwNL9EylxMpYTKACshPBHQ5EoTADOUdqDBMcxrDsBxnDcDwvB8PwAiCEIwgiODQExbFWOJfQrHYUwWG4SCtA7PJMoAHRAKUdy2VQNEOv4AF5+wOo6OsIU71EwQ7Jz+DiADkU07PbDroCAnAu67Yluv6AZAZ9wB7aUgP2FBVvWzbwH9TAXTdZBDEMFwLFMUwbDYFgzEMLbG3LStQS0AAVHp+1OSqoAQfsSd4-t2yCpg-hGChYj+HLNiDJwcLw+m-OqR9XrpP42DqlwZZlzK-ipiX7K8zqiEZJhMEZ8MmxbEg-g1-WIEUPI-jINQnDmdiLL1QSGT+cNNeODmuZFh8n0y9Td2JNhbQePYlv0lAcbYKx8a2tT3l2274FiE6zuekArpuw6Y7jp6Xrez7vpB-7AeT7Rc-B8Fu0laHZRQGwXBYKxce4YcUcDULMex3H8cJ4ntdJtzm0p6ndhafD7M75nu4gP42fkznSG53n0mvQXB-vMW-glqXZZlqx5cVxllfs1WDP1vumd1oMDeOY3TZWC3uZja2rOPRkHcN53p9dsWPbaL24bWtgNt9xacR4mQDYFgFgwHh3VFHQ6AAvOA5w87AxgXAjOH0vpQOgkZBBt1OprhQVndBx0HrxywYdQhj1zrg0zmgn6Yo4CxyIenROQNo50LThQiGJcAKbBhgcZAxh9DGHxlYOuyNUZQCbljHGeMCbmA7mWXiZMe7vCpprfudMGZD3kS2Ue49covxnvmfmC9haQWXqvaWstN4K0YErSie90gHwNponWQY9ZnyNkiS+5tLa30skeGy9tSCOzHlPbmpj3ae39BXKuNcbArmkEAA). The blocking chart looks fine in the Vega Editor. The Altair chart saved as a PNG image when rendered to PDF is not displayed correctly. The PDF-generated PNG image has empty white space to the left side of the y-title to fill out the whole width of the display, as shown above.  

Update 3: I am using Quarto 1.4.550 (latest) on Windows 10. Since the Altair chart is correctly displayed in Vega editor but the saved PNG is not correctly displayed (regardless of Quarto ), I conclude that it is an Altair issue: **Altair does not correctly `chart.save('chart.png')` the Splink blocking chart.**

Update 4: I concluded that the issue primarily is with Altair. Altair is not correctly saving the Splink blocking comparison chart as PNG. Therefore, I have now opened [Altair discussion issue 3336](https://github.com/altair-viz/altair/discussions/3336).

Update 5: Altair has now raised [vl-convert issue 152](https://github.com/vega/vl-convert/issues/152) for this.

Update 6: The bug has been fixed in [vega/vl-convert/pull/153](https://github.com/vega/vl-convert/pull/153). The bug fix will be available in the next `vl-convert` release 1.2.4.

Update 7: The corrected [vl-convert 1.2.4](https://github.com/vega/vl-convert/releases/tag/v1.2.4) has been released.

**Comment by RobinL** [comment, marked as answer]
Hiya.  Thanks for raising this - was just about to look into it but pleased to find you now have a solution!  Thanks for going the extra mile, getting it reported and fixed.- great work!!

**Reply by aalexandersson** [reply]
Thank you so much, very glad to hear this.
